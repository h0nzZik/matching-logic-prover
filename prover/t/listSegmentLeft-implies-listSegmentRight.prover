imports "t/definitions/listSegmentLeft.prover"
imports "t/definitions/listSegmentRight.prover"

claim \implies( \and( listSegmentLeft (variable("H") { ArrayIntInt } , variable("X") { Int }, variable("Y") { Int }, variable("F") { SetInt }) )
              , \and( listSegmentRight(variable("H") { ArrayIntInt } , variable("X") { Int }, variable("Y") { Int }, variable("F") { SetInt }) )
              )

// strategy search-bound(5)

strategy kt ; normalize ; ( ( simplify ; right-unfold-Nth(0, 0) ; smt )
                          | ( kt-solve-implications(smt) ; normalize
                            ; kt ; ( ( right-unfold-Nth(0, 1)
                                     ; right-unfold-Nth(0, 0)
                                     ; normalize
                                     ; instantiate-existentials
                                     ; smt
                                     )
                                   | ( right-unfold-Nth(0, 1)
                                     ; instantiate-existentials
                                     ; instantiate-universals-with-ground-terms
                                     ; normalize
                                     ; smt
                                     )
                                   )
                            )
                          )
