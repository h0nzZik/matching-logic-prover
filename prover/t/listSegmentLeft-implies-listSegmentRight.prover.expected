<prover>
  <k>
    .
  </k>
  <goals>
    <goal>
      <id>
        root
      </id>
      <active>
        true
      </active>
      <parent>
        .
      </parent>
      <claim>
        \implies ( \and ( listSegmentLeft ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "Y" ) { Int } , variable ( "F" ) { SetInt } , .Patterns ) , .Patterns ) , \exists { .Patterns } \and ( listSegmentRight ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "Y" ) { Int } , variable ( "F" ) { SetInt } , .Patterns ) , .Patterns ) )
      </claim>
      <strategy>
        success
      </strategy>
      <trace>
        .
      </trace>
    </goal>
  </goals>
  <declarations>
    <declaration>
      axiom \forall { variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "Y" ) { Int } , variable ( "F" ) { SetInt } , .Patterns } \iff-lfp ( listSegmentLeft ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "Y" ) { Int } , variable ( "F" ) { SetInt } , .Patterns ) , \or ( \and ( \equals ( variable ( "X" ) { Int } , variable ( "Y" ) { Int } ) , \equals ( variable ( "F" ) { SetInt } , emptyset ) , .Patterns ) , \exists { variable ( "X1" ) { Int } , variable ( "F1" ) { SetInt } , .Patterns } \and ( gt ( variable ( "X" ) { Int } , 0 , .Patterns ) , listSegmentLeft ( variable ( "H" ) { ArrayIntInt } , variable ( "X1" ) { Int } , variable ( "Y" ) { Int } , variable ( "F1" ) { SetInt } , .Patterns ) , \equals ( select ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , .Patterns ) , variable ( "X1" ) { Int } ) , \equals ( variable ( "F" ) { SetInt } , union ( variable ( "F1" ) { SetInt } , singleton ( variable ( "X" ) { Int } , .Patterns ) , .Patterns ) ) , disjoint ( variable ( "F1" ) { SetInt } , singleton ( variable ( "X" ) { Int } , .Patterns ) , .Patterns ) , .Patterns ) , .Patterns ) )
    </declaration> <declaration>
      axiom \forall { variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "Y" ) { Int } , variable ( "F" ) { SetInt } , .Patterns } \iff-lfp ( listSegmentRight ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "Y" ) { Int } , variable ( "F" ) { SetInt } , .Patterns ) , \or ( \exists { .Patterns } \and ( \equals ( variable ( "X" ) { Int } , variable ( "Y" ) { Int } ) , \equals ( variable ( "F" ) { SetInt } , emptyset ) , .Patterns ) , \exists { variable ( "Y1" ) { Int } , variable ( "F1" ) { SetInt } , .Patterns } \and ( listSegmentRight ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "Y1" ) { Int } , variable ( "F1" ) { SetInt } , .Patterns ) , gt ( variable ( "Y1" ) { Int } , 0 , .Patterns ) , \equals ( variable ( "Y" ) { Int } , select ( variable ( "H" ) { ArrayIntInt } , variable ( "Y1" ) { Int } , .Patterns ) ) , \equals ( variable ( "F" ) { SetInt } , union ( variable ( "F1" ) { SetInt } , singleton ( variable ( "Y1" ) { Int } , .Patterns ) , .Patterns ) ) , disjoint ( variable ( "F1" ) { SetInt } , singleton ( variable ( "Y1" ) { Int } , .Patterns ) , .Patterns ) , .Patterns ) , .Patterns ) )
    </declaration> <declaration>
      symbol listSegmentLeft ( ArrayIntInt , Int , Int , SetInt , .Sorts ) : Bool
    </declaration> <declaration>
      symbol listSegmentRight ( ArrayIntInt , Int , Int , SetInt , .Sorts ) : Bool
    </declaration>
  </declarations>
</prover>
